packages/flutter_tools/lib/src/build_runner/build_runner.dart

```diff
@@ -136,12 +150,146 @@ class BuildRunner {
     return BuildResult(packagesFile, dillFile);
   }
 
-  void _handleOutput(String line) {
-    printTrace(line);
+  /// Invalidates a generated build script by deleting it.
+  ///
+  /// Must be called any time a pubspec file update triggers a corresponding change
+  /// in .packages.
+  Future<void> invalidateBuildScript() async {
+    final FlutterProject flutterProject = await FlutterProject.current();
+    final File buildScript = flutterProject.dartTool
+      .absolute
+      .childDirectory('flutter_tool')
+      .childFile('build.dart');
+    if (!await buildScript.exists()) {
+      return;
+    }
+    await buildScript.delete();
+  }
+
+  // Generates a synthetic package script under .dart_tool/flutter_tool.
+  // This is in turn used to generate a build script, which is coppied to the
+  // correct location in the original package and patched to correctly refer
+  // to the synthetic .packages file and pubspec.
+  Future<void> generateBuildScript() async {
+    final FlutterProject flutterProject = await FlutterProject.current();
+    final String generatedDirectory = fs.path.join(flutterProject.dartTool.path, 'flutter_tool');
```

Comment:
Or `flutterProject.dartTool.childDirectory('flutter_tool').path`

Reply:
Seems like it might clean things up if you store `File` and `Directory` instances here, then refer to `path` as needed below?